Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/BaseSyncTool.kt

```diff
@@ -0,0 +1,99 @@
+package de.rki.coronawarnapp.diagnosiskeys.download
+
+import de.rki.coronawarnapp.appconfig.KeyDownloadConfig
+import de.rki.coronawarnapp.diagnosiskeys.server.LocationCode
+import de.rki.coronawarnapp.diagnosiskeys.storage.CachedKey
+import de.rki.coronawarnapp.diagnosiskeys.storage.CachedKeyInfo
+import de.rki.coronawarnapp.diagnosiskeys.storage.KeyCacheRepository
+import de.rki.coronawarnapp.storage.DeviceStorage
+import timber.log.Timber
+
+open class BaseSyncTool(
+    private val keyCache: KeyCacheRepository,
+    private val deviceStorage: DeviceStorage,
+    private val tag: String
+) {
+
+    internal suspend fun invalidateCachedKeys(invalidatedKeyFiles: Collection<KeyDownloadConfig.InvalidatedKeyFile>) {
+        if (invalidatedKeyFiles.isEmpty()) {
+            Timber.tag(tag).d("No invalid files to delete.")
+            return
+        }
+
+        val badEtags = invalidatedKeyFiles.map { it.etag }
+        val toDelete = keyCache.getAllCachedKeys()
+            .filter { badEtags.contains(it.info.etag) }
+
+        Timber.tag(tag).w("Deleting invalidated cached keys: %s", toDelete.joinToString("\n"))
+        keyCache.delete(toDelete.map { it.info })
```

Comment:
These files should not be deleted yet. They should be downloaded again and if the download was successful, overwrite the cached file.

Reply:
Why not?
Either they are invalid or they are not, why would we want to resubmit files that we know are invalid? I think it's better to leave out a file than to submit an invalid one. For all we know, the invalid file could break the ENF. If the download that is supposed to overwrite the file fails for some reasons, then the invalid file is still there.

Reply:
This is not only for invalid files that break ENF. Let's me try to explain...

If there's a rogue file on CDN that breaks ENF and that has already been downloaded by the client, the client might shows unexpected behavior (whatever that means in detail).
1. with the current implementation (delete before download)
    1. if the download fails, the client has one file less on the device and would or would not run Exposure Detection and Risk Calculation (depending on whether it's an hour our day package). If it runs, ENF should not fail anymore.
    1. if the download succeeds, everything is fine.
1. with the expected implementation (replace with new download instead of delete before download)
    1. if the download fails, the client still has the old file. Exposure Detection and Risk Calculation might _still_ fail 
    1. if the download succeeds, everything is fine.

But there's also the scenario where there's just a (maybe just partially) bad file that doesn't break ENF but leads to wrong results:
1. with the current implementation (delete before download)
    1. if the download fails, the client has one file less on the device and would or would not run Exposure Detection and Risk Calculation (depending on whether it's an hour our day package). If it runs, ENF might return a different number of matches. This might be lower for the user if a file is missing. This might falsly hide a red card because we removed the data too early.
    1. if the download succeeds, everything is fine.
1. with the expected implementation (replace with new download instead of delete before download)
    1. if the download fails, the client still has the old file. Exposure Detection and Risk Calculation would could still show the incorrect result but IMO that's better than no result (if the missing file leads to result)
    1. if the download succeeds, everything is fine.

So in the end, I'd rather have clients continue to fail than not showing a risk although there is one.